File: examples/fundamentals-java/src/main/java/example/enums/Cards.java
package example.enums;

/** Classifies a card suit by colour — using {@code ==}, not {@code switch}. */
public final class Cards {

    private Cards() {
    }

    public static boolean isRed(Suit s) {
        return s == Suit.HEARTS || s == Suit.DIAMONDS;
    }

    /** BUG: forgot SPADES — so SPADES is classified as neither red nor black. */
    public static boolean isBlack(Suit s) {
        return s == Suit.CLUBS;
    }

    /** The fix. */
    public static boolean isBlackFixed(Suit s) {
        return s == Suit.CLUBS || s == Suit.SPADES;
    }

    /** A {@code switch} over the enum — JBMC handles this soundly (see the proof). */
    public static int rank(Suit s) {
        switch (s) {
            case HEARTS: return 1;
            case DIAMONDS: return 2;
            case CLUBS: return 3;
            case SPADES: return 4;
            default: return 0;
        }
    }
}

File: examples/fundamentals-kotlin/src/main/kotlin/example/enums/Cards.kt
package example.enums

/** Classifies a card suit by colour, using `when` over the enum (sound under JBMC). */
object Cards {

    fun isRed(s: Suit): Boolean = when (s) {
        Suit.HEARTS, Suit.DIAMONDS -> true
        else -> false
    }

    /** BUG: forgot SPADES — so SPADES is classified as neither red nor black. */
    fun isBlack(s: Suit): Boolean = when (s) {
        Suit.CLUBS -> true
        else -> false
    }

    /** The fix. */
    fun isBlackFixed(s: Suit): Boolean = when (s) {
        Suit.CLUBS, Suit.SPADES -> true
        else -> false
    }

    /** An exhaustive `when` over the enum — JBMC handles it soundly (see the proof). */
    fun rank(s: Suit): Int = when (s) {
        Suit.HEARTS -> 1
        Suit.DIAMONDS -> 2
        Suit.CLUBS -> 3
        Suit.SPADES -> 4
    }
}
